### PR TITLE
Process multiple OCSP responses

### PR DESCRIPTION
**Pull request by tmael**
ZD#10987.

support for OCSP request/response for multiple certificates

Test case: https://github.com/wolfSSL/wolfssl-examples/pull/231

### CODE REVIEW

File: wolfcrypt/src/asn.c [hunk 1]
```diff
@@ -16546,115 +16547,120 @@ static int DecodeSingleResponse(byte* source,
...
-     * at a time. There should only be one SingleResponse */
+    /* wolfSSL only requests one status for one certificate at a time but
+       some OCSP responders can reply with multiple SingleResponse items.
+       Expect to handle one SingleResponse. Otherwise, we can process the
+       responses but only the last entry in the list is verified. */
 
-    /* Wrapper around the Single Response */
-    if (GetSequence(source, &idx, &length, size) < 0)
-        return ASN_PARSE_E;
+    while ((idx-prevIndex) < (word32)wrapperSz) {
```
Comment (dgarske) [inline wolfcrypt/src/asn.c:16555]: This while loop is the only addition here (rest is indent). While it properly iterates the responses it's only capturing the last "singleResponse" into `resp`. Either you need to filter based on the request data or you'll need to capture a list of these to compare later. Specifically `issuerHash`, `issuerKeyHash`, `serial`, `serialSz`, `status`, `thisDate`, `nextDate`.

In the customer report it looks like the responses are all for the same key, but different serial numbers.

**Review by dgarske — CHANGES_REQUESTED** [review]

File: wolfcrypt/src/asn.c [hunk 2]
```diff
@@ -16806,8 +16781,26 @@ static int DecodeResponseData(byte* source,
...
+    if (GetSequence(source, &idx, &wrapperSz, size) < 0)
+        return ASN_PARSE_E;
+
+    localIdx = idx;
+    cs = resp->status;
+    while (idx - localIdx < (word32)wrapperSz) {
+        if ((ret = DecodeSingleResponse(source, &idx, resp, localIdx + wrapperSz, cs)) < 0)
+            return ret; /* ASN_PARSE_E, ASN_BEFORE_DATE_E, ASN_AFTER_DATE_E */
+        if (idx - localIdx < (word32)wrapperSz) {
+            cs->next = (CertStatus*)XMALLOC(sizeof(CertStatus), resp->heap,
```
Comment (tmael) [inline wolfcrypt/src/asn.c:16798]: Does it make sense to limit the number of OCSP status responses we can process?

File: wolfcrypt/src/asn.c (hunk 1 above)
Comment (tmael) [reply wolfcrypt/src/asn.c]: @dgarske. You are spot on! expanded to process the OCSP responses only and caveats were documented as a comment.

**Comment by tmael** [comment]
Thanks, @dgarske, Nice work! 
No issues with the following test configurations. 
-  client and server are able to connect with an OCSP responder with the id-pkix-ocsp-nocheck cert extensions.  
    - server makes OCSP request and processes the OCSP response from the responder. 
    - client skips performing OCSP request/response with the OCSP responder. 
- Verified OcspResponseDecode() can now process multiple SingleResponses OCSP responses and stores it in a CertStatus link list.

File: wolfcrypt/src/asn.c [hunk 3]
```diff
@@ -17310,13 +17315,27 @@ int CompareOcspReqResp(OcspRequest* req, OcspResponse* resp)
         return cmp;
     }
 
-    cmp = req->serialSz - resp->status->serialSz;
-    if (cmp != 0) {
-        WOLFSSL_MSG("\tserialSz mismatch");
-        return cmp;
+    /* match based on found status and return */
+    for (status = resp->status; status; status = next) {
```
Comment (tmael) [inline wolfcrypt/src/asn.c:17323]: The function is good for our current use case where we have the same _issuerNameHash_ and  _issuerKeyHash_ but different _serialNumber_. 

Does it make sense to move the for loop at the beginning of the function to handle a general case?

File: wolfcrypt/src/asn.c [hunk 4]
```diff
@@ -16991,6 +16984,17 @@ void InitOcspResponse(OcspResponse* resp, CertStatus* status,
     resp->status         = status;
     resp->source         = source;
     resp->maxIdx         = inSz;
+    resp->heap           = heap;
+}
+
+void FreeOcspResponse(OcspResponse* resp)
+{
```
Comment (tmael) [inline wolfcrypt/src/asn.c:16995]: check resp != NULL

File: wolfcrypt/src/asn.c (hunk 3 above)
Comment (dgarske) [reply wolfcrypt/src/asn.c:17323]: `CertStatus` does not contain `issuerHash` or `issuerKeyHash`. The OCSP responses are all for the same key, but different serial numbers. Serial number is all we can match on here.

File: wolfcrypt/src/asn.c (hunk 4 above)
Comment (dgarske) [reply wolfcrypt/src/asn.c:16995]: This is an internal function and the other OCSP API's also do not check args. Seems like a good future cleanup operation...

**Review by dgarske — APPROVED** [review]
Todd should do a final on this. Both Tesfa and I worked on the PR.

**Review by toddouska — APPROVED** [review]

**Pull request merged**